specs/plasma.md

```diff
@@ -0,0 +1,114 @@
+`challenge_window` and `resolve_window` are constant values that currently cannot be changed
+unless the contract is upgraded. A dynamic window mechanism may be explored in the future.
+
+## Derivation
+
+Input data is retrieved during derivation of L2 blocks. The changes to the derivation pipeline
+when using the alt DA source are limited to swapping out the L1 based `DataAvailabilitySource` step
+in the pipeline with a module that enables pulling the data from the offchain DA source.
+
+Similarly to L1 based DA, for each L1 block we open a calldata source to retrieve the input commitments
+from the transactions and use each commitment with its l1 origin block number to resolve
+the input data from the storage service.
+
+In addition, we filter events from the DA Challenge contract included in the block
+and sync a local state of challenged input commitments. As the derivation pipeline steps through
+`challenge_window + resolve_window` amount of L1 blocks, any challenge marked as active
+becomes expired causing a reset of the derivation pipeline.
+
+Derivation can either be driven by new L1 blocks or restarted (reset) from the L1 origin of the last
+L2 safe head known by the execution engine.
+
+When stepping through new L1 origin blocks, input data is loaded from the DA storage service.
+If the service responds with a 404 (not found) error, derivation stalls and the DA Manager
+starts a detached traversal of new L1 blocks until:
+
+- An `Active` challenge event is found, detached L1 traversal continues:
+  - A `Resolved` challenge event is found: derivation continues with the input data submitted onchain.
+  - A high enough L1 block is reached such as `latest_l1 - pipeline_l1_origin = resolve_window_size`:
+    input data is skipped from the derivation output.
+- Data is never challenged, derivation returns critical error.
+
+When derivation is reset, the pipeline steps through previously seen L1 origins so that
+`block_start = latest_l1_block - (challenge_window_size + resolve_window_size + sequencer_window_size)`.
+In that case, the DA manager has already synced the state of challenges during the range of L1 blocks traversed
+so the pipeline can skip commitments with expired challenges and reorg the L2 chain
+or load input data from the resolving transaction calldata.
+
+## Safety and finality
+
+Similarly to rollup mode, the engine queue labels any new blocks derived from input data with a commitment
+on the L1 chain as “safe”. Although labeled as “safe”, the chain might still reorg in case of a faulty DA provider
+and users must use the “finalized” label for a guarantee that their state cannot revert.
+
+With Plasma mode on, the engine queue does not receive finality signals from the L1 RPC
+but from the DA manager that keeps track of challenges. The engine queue will maintain a longer buffer
+of L2 blocks waiting for the DA windows to expire in order to be finalized.
```

Comment:
We should ensure that the finalized head only updates when both the L1 block the commitment is in is finalized and the DA windows have expired. Normally the DA window is the longer time period, but the L1 chain may be in a long period of non-finality such that the DA window expires before the commitment is finalized.

Reply:
Finality is not always 64 blocks but can be delayed during periods where not enough nodes are active.